Add support for the `format` property in query

carolinan opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

Adds handling the format property ($request['format'] ) in the class WP_REST_Posts_Controller.
Adds handling the format property ($block->context['query']['format'])  in the function build_query_vars_from_query_block.

Trac ticket: https://core.trac.wordpress.org/ticket/62014


---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

carolinan committed:
Add support for the `format` property in query

Adds handling the format property ($request['format'] ) in the class WP_REST_Posts_Controller.

Adds handling the format property ($block->context['query']['format'])  in the function build_query_vars_from_query_block.

carolinan committed:
Tests: Add format as a valid parameter in WP_Test_REST_Posts_Controller::test_registered_query_params

carolinan committed:
build_query_vars_from_query_block: Fix incorrect variable name and broken logic

In `build_query_vars_from_query_block`, the variable that contains the format attribute had the wrong name, which caused a PHP warning. The variable name was updated to `$formats`.

One of the if-statements was closed incorrectly, which meant that all the logic for handling the format property was not used. Because of this, the wrong posts showed in the query loop block on the front of the site.
The closing curly bracket for the if-statement was moved to solve this.

One additional check was added to the wrapping if-statement:
It now checks that `$block->context['query']['format']` is not empty, and is an array, before running the rest of the code.

carolinan committed:
Tests: Update the qunit fixture in `wp-api-generated.js`

Generated an updated fixture in `wp-api-generated.js` by running the PHPUnit tests.
The update adds `format` to `mockedApiResponse.Schema`

carolinan committed:
`WP_REST_Posts_Controller`: remove redundant array conversion for the post formats.

`WP_REST_Posts_Controller`: remove unnecessary array conversion for the `format` parameter.
This addresses feedback that was left on the Gutenberg pull request.

carolinan committed:
`WP_REST_Posts_Controller`: Tidy up the inline comments.

carolinan committed:
`WP_REST_Posts_Controller` `get_collection_params`: add `uniqueItems` to `$query_params['format']`

In `WP_REST_Posts_Controller` `get_collection_params` class method; add `uniqueItems` to the `$query_params['format']` array.

The purpose is to require all items in the array to be unique.
This will help manage user errors, if duplicate items are passed in the `format` parameter.

This change addresses feedback left on the Gutenberg pull request.

carolinan committed:
Tests: Update the qunit fixtures in wp-api-generated.js

Adds the regenerated fixtures which includes `uniqueItems` in the `format` array in `mockedApiResponse.Schema`.

The fixture was regenerated by running the PHPUNit tests.

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,48 @@ function build_query_vars_from_query_block( $block, $page ) {
+				unset( $formats[ array_search( 'standard', $formats, true ) ] );
+			}
+			// Add any remaining formats to the tax query.
+			if ( ! empty( $formats ) ) {
```
`! empty( $block->context['query']['format'] )` is already checked in line no. 2424. Should be removed.

Mamaduka replied:
This is a check after the `standard` format is removed. See L2439.

carolinan replied:
It is intended the check if there are any values left in case the 'standard' value was removed.
There is no need to run this code, if standard was the only value.

mukeshpanchal27 replied:
Got it. Thanks!

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,48 @@ function build_query_vars_from_query_block( $block, $page ) {
+			$formats   = $block->context['query']['format'];
+			$tax_query = array( 'relation' => 'OR' );
+
+			// The default post format, `standard`, is not stored in the database.
+			// If `standard` is part of the request, the query needs to exclude all post items that
+			// have a format assigned.
```
```suggestion
			/*
			 * The default post format, `standard`, is not stored in the database.
			 * If `standard` is part of the request, the query needs to exclude all post items that
			 * have a format assigned.
			 */
```
Use multi-line comment.

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,48 @@ function build_query_vars_from_query_block( $block, $page ) {
+			if ( ! isset( $query['tax_query'] ) ) {
+				$query['tax_query'] = array();
+			}
+			// This condition is intended to prevent `$tax_query` from being added to `$query`
+			// if it only contains the relation.
```
```suggestion
			/*
			 * This condition is intended to prevent `$tax_query` from being added to `$query`
			 * if it only contains the relation.
			 */
```

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -339,6 +339,53 @@ public function get_items( $request ) {
+			}
+
+			// Add any remaining formats to the tax query.
+			if ( ! empty( $formats ) ) {
```
Duplicate condition.

Mamaduka replied:
Same as https://github.com/WordPress/wordpress-develop/pull/7314/files#r1754306934.

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -339,6 +339,53 @@ public function get_items( $request ) {
+			$formats   = $request['format'];
+			$tax_query = array( 'relation' => 'OR' );
+
+			// The default post format, `standard`, is not stored in the database.
+			// If `standard` is part of the request, the query needs to exclude all post items that
+			// have a format assigned.
```
```suggestion
			/*
			 * The default post format, `standard`, is not stored in the database.
			 * If `standard` is part of the request, the query needs to exclude all post items that
			 * have a format assigned.
			 */
```

carolinan committed:
Update the format of inline comments

Use multiline comments instead of multiple single line comments.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,52 @@ function build_query_vars_from_query_block( $block, $page ) {
+				$tax_query[] = array(
+					'taxonomy' => 'post_format',
+					'field'    => 'slug',
+					'terms'    => array(),
```
Nit: This shouldn't be needed with the operator NOT EXISTS.

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,52 @@ function build_query_vars_from_query_block( $block, $page ) {
 			}
 		}
+		if ( ! empty( $block->context['query']['format'] ) && is_array( $block->context['query']['format'] ) ) {
+			$formats   = $block->context['query']['format'];
```
Validation: Ensure that the formats come from the supported values only provided by `get_post_format_slugs()`.

⚠️ Untested but I think you want something along the lines of:

```php
$formats = array_intersect( $formats, get_post_format_slugs() );
```

carolinan replied:
Though could I add this after the check for the standard format, since format is not (usually) in get_post_format_slugs()?

peterwilsoncc commented on src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php:
```diff
@@ -339,6 +339,55 @@ public function get_items( $request ) {
+				$tax_query[] = array(
+					'taxonomy' => 'post_format',
+					'field'    => 'slug',
+					'terms'    => array(),
```
As above

peterwilsoncc commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,52 @@ function build_query_vars_from_query_block( $block, $page ) {
 		}
+		if ( ! empty( $block->context['query']['format'] ) && is_array( $block->context['query']['format'] ) ) {
+			$formats   = $block->context['query']['format'];
+			$tax_query = array( 'relation' => 'OR' );
```
My reading of this code is that if someone sets up a query for both formats and taxomomies, this will return posts that contain either.

Eg a query for `post format: link; category: rest-api` will return:
* Any posts with the post format link, regardless of category
* Any posts in the category rest-api, regardless of format.

carolinan replied:
Yeah I think something is not working correctly.
If I use `OR` or `AND`, the result is the same: Posts that have **both** the category and the format. 
I think the correct use is `AND`.
But then why is not `OR` returning all posts that have either the format or the category?

carolinan replied:
Here is an example query loop for listing posts in category 4, with either the gallery or link format. The category id needs to be updated to match a category available on the installation.

```
<!-- wp:query {"queryId":1,"query":{"perPage":10,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":false,"format":["link","gallery"],"taxQuery":{"category":[4]}}} -->
<div class="wp-block-query"><!-- wp:post-template -->
<!-- wp:post-title /-->
<!-- /wp:post-template --></div>
<!-- /wp:query -->
```

carolinan replied:
I have changed it to `AND` though it bothers me why `OR` is working the way I am seeing it in testing 😄

carolinan replied:
🤔 In the `WP_REST_Posts_Controller` class, there is also:

```
// Enable filtering by both post formats and other taxonomies by combining them with `AND`.
			if ( isset( $args['tax_query'] ) ) {
				$args['tax_query'][] = array(
					'relation' => 'AND',
					$tax_query,
				);
			} else {
				$args['tax_query'] = $tax_query;
			}
```
https://github.com/WordPress/wordpress-develop/pull/7314/files#diff-48cfdad3b747164ac4b6857186399519dabf913026235575b00ebe5b0dc7169bR383

carolinan replied:
I am badly regretting not documenting why I added this relation in the first place.

Once I get back to work I will continue doing more logging of the different queries:
But I am now thinking that I meant for this to be the relation **between the formats**.
That I choose 'OR' because the user can query for both 'video' and 'link' and it needs to be 'OR' because a post can only have one format. It always have to list posts that has either 'video' or 'link', never both.

carolinan replied:
Also I think `$tax_query` was not a good name for the variable, it too can cause confusion.

carolinan replied:
After more testing and logging of `$args['tax_query']` in the `WP_REST_Posts_Controller` class, I found that:

- Setting the relation on the array to either `OR` or `AND` works when the query has two formats such as link and gallery. Because there is only one condition; the relation is not actually used, there are no conditions to create a relation between.

- Using `OR` is needed when the query includes `standard` **and** a non-default format such as link and gallery, because now there are two conditions.

So that is why the testing earlier showed the same results: I was not testing with a combination of `standard`.

peterwilsoncc reviewed (commented):
I've added some note inline, the most important being for validating the values when generating the query within the block. I'll leave it up to you to decide on how to handle invalid data.

It would also be good to see some unit tests added to ensure that the REST endpoint is returning an expected set of posts. Particularly for `standard` as the need to query for a non-existent taxomomy is a little unusual.

carolinan committed:
Remove unnecessary `terms` from the query for the standard format

In the `build_query_vars_from_query_block` function and the `WP_REST_Posts_Controller` class, a query with the `NOT EXISTS` operator is used to exclude all items that have a post format assigned.

This commit removes a `terms` parameter from the two queries, because the terms are redundant when `NOT EXISTS` is used.

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,51 @@ function build_query_vars_from_query_block( $block, $page ) {
+				// Add the `post-format-` prefix.
+				$terms = array_map(
+					static function ( $format ) {
+						return 'post-format-' . $format;
```
```suggestion
						return "post-format-$format";
```

Nit.

carolinan committed:
Change the `relation` from `OR` to `AND` when querying for the `format`

When querying for the `format` property in the `build_query_vars_from_query_block` function and the `WP_REST_Posts_Controller` class, the query may be combined with other taxonomy queries. For example, the user may expect the result to be posts with a specific format and a specific category.

Previously, the `relation` parameter in the query was set to `OR`, which should return items that has either the format or the taxonomy.

This commit changes the `relation` to `AND` with the intention to limit the result to items that have both the queried taxonomy and the format.

carolinan committed:
In build_query_vars_from_query_block: validate the formats

Validate the that the `format` property passed to the query using the block context is either the default (standard) or a supported post format, registered with `add_theme_support()`.

Compare `format` to the supported formats in  `get_post_format_slugs()` and remove invalid formats.

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -2420,6 +2421,63 @@ function build_query_vars_from_query_block( $block, $page ) {
+			 * - First, add `standard` to the array of valid formats.
+			 * - Then, remove any invalid formats.
+			 */
+			$valid_formats = array_merge( ['standard'], get_post_format_slugs() );
```
```suggestion
			$valid_formats = array_merge( array( 'standard' ), get_post_format_slugs() );
```

PHPCS fix

carolinan committed:
WP_REST_Posts_Controller: Change the `relation` from `AND` to `OR` when querying for the `format`

When querying for the `format` property in the `WP_REST_Posts_Controller` class,  the `relation` parameter in the query was incorrectly set to `AND`.
This commit changes the relation back to `OR`.

This is needed because when a user queries for a combination of the `standard` post format and any other format, the result should include items with either the given format, or no format (standard).
A post item can not have more than one format.

The commit also renames the variable `$tax_query` to `$formats_query` to improve the readability.

carolinan committed:
build_query_vars_from_query_block: Change the `relation` when querying for the `format`

When querying for the `format` property in `build_query_vars_from_query_block`, the `relation` parameter in the query was incorrectly set to `AND`.
This commit changes the relation back to `OR`.

This is needed because when a user queries for a combination of the `standard` post format and any other format, the result should include items with either the given format, or no format (standard).
A post item can not have more than one format.

The commit also renames the variable `$tax_query` to `$formats_query` to improve the readability.

carolinan committed:
Formats: address feedback from code review and PHPCS issues

This commit corrects inline docBlocks, removes unnecessary string concatenations, and tries to resolve a PHP coding standard issue about the formatting of an array.

carolinan committed:
build_query_vars_from_query_block: Try to resolve the nested relations

When querying for formats, the relation between the format and other taxonomies like categoryíes and tags should be `AND`.

The intention is to allow users to query for a post with a specific category and a specific format.

noisysocks commented:
Hey @carolinan. JavaScript packages were updated in r59072 so you might wish to rebase this for easier testing. The deadline to commit this backport is 6.7 Beta 1 which is scheduled for 1 October.

carolinan committed:
Merge branch 'WordPress:trunk' into try/post-formats-62014

carolinan committed:
Merge branch 'WordPress:trunk' into try/post-formats-62014

peterwilsoncc commented:
I tested this with the updated packages and was getting a block error when I tried to add the format. Is there something that needs to be merged to the packages in the GB repo before it should work? 

![query-loop](https://github.com/user-attachments/assets/05e8f65e-0adc-4239-ba57-4540ed51c3dd)



It's a little confusing with the same variable name in both the legacy query, the tax query so I'm wondering if we can do a little more tidying up there:

```diff
diff --git a/src/wp-includes/blocks.php b/src/wp-includes/blocks.php
index f3d138475f..25cb1c0d24 100644
--- a/src/wp-includes/blocks.php
+++ b/src/wp-includes/blocks.php
@@ -2346,6 +2346,7 @@ function build_query_vars_from_query_block( $block, $page ) {
 		'order'        => 'DESC',
 		'orderby'      => 'date',
 		'post__not_in' => array(),
+		'tax_query'    => array(),
 	);
 
 	if ( isset( $block->context['query'] ) ) {
@@ -2395,34 +2396,35 @@ function build_query_vars_from_query_block( $block, $page ) {
 		}
 		// Migrate `categoryIds` and `tagIds` to `tax_query` for backwards compatibility.
 		if ( ! empty( $block->context['query']['categoryIds'] ) || ! empty( $block->context['query']['tagIds'] ) ) {
-			$tax_query = array();
+			$tax_query_back_compat = array();
 			if ( ! empty( $block->context['query']['categoryIds'] ) ) {
-				$tax_query[] = array(
+				$tax_query_back_compat[] = array(
 					'taxonomy'         => 'category',
 					'terms'            => array_filter( array_map( 'intval', $block->context['query']['categoryIds'] ) ),
 					'include_children' => false,
 				);
 			}
 			if ( ! empty( $block->context['query']['tagIds'] ) ) {
-				$tax_query[] = array(
+				$tax_query_back_compat[] = array(
 					'taxonomy'         => 'post_tag',
 					'terms'            => array_filter( array_map( 'intval', $block->context['query']['tagIds'] ) ),
 					'include_children' => false,
 				);
 			}
-			$query['tax_query'] = $tax_query;
+			$query['tax_query'] = array_merge( $query['tax_query'], $tax_query_back_compat );
 		}
 		if ( ! empty( $block->context['query']['taxQuery'] ) ) {
-			$query['tax_query'] = array();
+			$tax_query = array();
 			foreach ( $block->context['query']['taxQuery'] as $taxonomy => $terms ) {
 				if ( is_taxonomy_viewable( $taxonomy ) && ! empty( $terms ) ) {
-					$query['tax_query'][] = array(
+					$tax_query[] = array(
 						'taxonomy'         => $taxonomy,
 						'terms'            => array_filter( array_map( 'intval', $terms ) ),
 						'include_children' => false,
 					);
 				}
 			}
+			$query['tax_query'] = array_merge( $query['tax_query'], $tax_query );
 		}
 		if ( ! empty( $block->context['query']['format'] ) && is_array( $block->context['query']['format'] ) ) {
 			$formats = $block->context['query']['format'];
@@ -2479,13 +2481,14 @@ function build_query_vars_from_query_block( $block, $page ) {
 			 */
 			if ( count( $formats_query ) > 1 ) {
 				// Enable filtering by both post formats and other taxonomies by combining them with `AND`.
-				if ( isset( $query['tax_query'] ) ) {
-					$query['tax_query'][] = array(
+				if ( empty( $query['tax_query'] ) ) {
+					$query['tax_query'] = $formats_query;
+				} else {
+					$query['tax_query'] = array(
 						'relation' => 'AND',
+						$query['tax_query'],
 						$formats_query,
 					);
-				} else {
-					$query['tax_query'] = $formats_query;
 				}
 			}
 		}
```

However, the code looks good to me in terms of PHP, it's just the blcok crashing that needs to be figured out.

carolinan commented:
I have submitted [a pull request](https://github.com/WordPress/gutenberg/pull/65551) that fixes this bug and it is waiting for review.

carolinan commented:
>  I'm wondering if we can do a little more tidying up there:

I don't mind :)

carolinan committed:
Merge branch 'WordPress:trunk' into try/post-formats-62014

carolinan commented:
@peterwilsoncc My only question is about the first `array_merge`. The previous code replaced `$query['tax_query']`, the suggested update merges them, I would like to understand why?
```
-			$query['tax_query'] = $tax_query;
+			$query['tax_query'] = array_merge( $query['tax_query'], $tax_query_back_compat );
```

carolinan committed:
build_query_vars_from_query_block: Update variable names and if/else logic

In function `build_query_vars_from_query_block`:
Address feedback from the code review.

- Update the variable name used for the tax query to improve readability.
Inside the condition that migrates category- and tag id's for backwards compatibility, the new variable name for the tax query is `$tax_query_back_compat`.

- Use `array_merge` to merge `$query['tax_query']` with the edited queries inside each condition.

-Update the logic of the if/else statement used to determine wether the `format` parameter should combined with an existing tax query.
The purpose of the update is to:
-Ensure that the `AND` relation between the taxonomies and the post formats is nested at the correct level in the array.
- Improve readability.

carolinan committed:
Oops

One change was not included in the previous commit.
This commit removes the left over `else` because it is incorrect.

carolinan committed:
Tests: Tests_Blocks_wpBlock: Update arrays used in `assertSame`

In `Tests_Blocks_wpBlock`, five test assertions were failing because the arrays used in the `assertSame` no longer matched the `$query` variable.

The missmatch was caused by an update to the function `build_query_vars_from_query_block`.
This commit updates the arrays in the test to match the function.

carolinan committed:
Merge branch 'WordPress:trunk' into try/post-formats-62014

peterwilsoncc committed:
Merge branch 'trunk' into try/post-formats-62014

peterwilsoncc commented:
> @peterwilsoncc My only question is about the first `array_merge`. The previous code replaced `$query['tax_query']`, the suggested update merges them, I would like to understand why?

The suggested change always defined `$query['tax_query']` and renamed the variables so they differed for the legacy queries, current query and formats. As a result it needed to merge rather than append the item. 

However, I'm glad you asked because I want to double check I didn't just introduce a bug. I'll double check that prior to merge.

peterwilsoncc committed:
Merge branch 'trunk' into try/post-formats-62014

peterwilsoncc committed:
Add tests for the rest api format getter.

peterwilsoncc committed:
Add tests for post formats in query block builder.

peterwilsoncc commented:
Merged [r59115](https://core.trac.wordpress.org/changeset/59115) / https://github.com/WordPress/wordpress-develop/commit/b1db74a765cd6dab13b7d2a09ef3890abcc59e25
